File: CPP Project/RSAHandler.cpp
#include "RSAHandler.h"
#include <cstdlib>
#include <cmath>
#include <numeric>

RSAHandler::RSAHandler(int key_size) {
    int64_t biggest_num = pow(2, key_size - 1);

    p = generate_prime(biggest_num);
    q = generate_prime(biggest_num);

    modulus = p * q;
    t = totient(modulus);

    publicKey = generate_e(t);
    privateKey = calculate_d(publicKey, t);
}

void RSAHandler::setClientPublicKey(int64_t clientPublicKey) {
    this->clientPublicKey = clientPublicKey;
}

void RSAHandler::setClientModulus(int64_t clientModulus) {
    this->clientModulus = clientModulus;
}

int64_t RSAHandler::getPublicKey() {
    return publicKey;
}

int64_t RSAHandler::getModulus() {
    return modulus;
}

int64_t RSAHandler::getPrivateKey() {
    return privateKey;
}

std::vector<int64_t> RSAHandler::encrypt(string m) {
    if (clientPublicKey == 0 || clientModulus == 0) {
        throw std::runtime_error("Client public key or modulus not set");
    }

    std::vector<int64_t> encryptedMessage;

    for (char c : m) {
        int64_t encryptedChar = mod_pow(c, clientPublicKey, clientModulus);
        encryptedMessage.push_back(encryptedChar);
    }

    return encryptedMessage;
}

std::string RSAHandler::decrypt(const std::vector<int64_t>& encrypted) {
    std::string decryptedMessage;

    for (int64_t encryptedChar : encrypted) {
        char decryptedChar = static_cast<char>(mod_pow(encryptedChar, privateKey, modulus));
        decryptedMessage += decryptedChar;
    }

    return decryptedMessage;
}

#include "RSAHandler.h"

int64_t RSAHandler::generate_prime(int64_t big_num) {
    int64_t prime = rand() % big_num + 1;

    while (!is_prime(prime)) {
        prime = rand() % big_num + 1;
    }

    return prime;
}

int64_t RSAHandler::totient(int64_t n) {
    int64_t result = n;

    for (int64_t i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) {
            while (n % i == 0) {
                n /= i;
            }

            result -= result / i;
        }
    }

    if (n > 1) {
        result -= result / n;
    }

    return result;
}

bool RSAHandler::is_prime(int64_t n) {
    if (n <= 1) return false;

    for (int64_t i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) return false;
    }

    return true;
}

int64_t RSAHandler::generate_e(int64_t num) {
    for (int64_t e = 2; e < num; e++) {
        if (gcd(e, num) == 1) {
            return e;
        }
    }

    throw std::runtime_error("Couldn't generate e");
}

int64_t RSAHandler::mod_pow(int64_t base, int64_t exponent, int64_t modulus) {
    int64_t result = 1;

    while (exponent > 0) {
        if (exponent % 2 == 1) {
            result = (result * base) % modulus;
        }
        base = (base * base) % modulus;
        exponent /= 2;
    }

    return result;
}

int64_t RSAHandler::calculate_d(int64_t e, int64_t totient) {
    int64_t d = mod_inverse(e, totient);
    return d;
}

int64_t RSAHandler::mod_inverse(int64_t a, int64_t m) {
    int64_t m0 = m;
    int64_t y = 0, x = 1;

    if (m == 1) return 0;

    while (a > 1) {
        int64_t q = a / m;
        int64_t t = m;

        m = a % m;
        a = t;
        t = y;

        y = x - q * y;
        x = t;
    }

    if (x < 0) x += m0;

    return x;
}

int64_t RSAHandler::gcd(int64_t a, int64_t b) {
    while (b != 0) {
        int64_t temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

File: MappingServer/src/RSAHandler.kt
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sqrt


class RSAHandler(keySize: Int = 2048 * 2 * 2) {
    private var p: ULong = 0UL
    private var q: ULong = 0UL
    private var modulus: ULong = 0UL
    private var t: ULong = 0UL
    private var publicKey: ULong = 0UL
    private var privateKey: ULong = 0UL
    private var clientPublicKey: ULong = 0UL
    private var clientModulus: ULong = 0UL

    init {
        p = generatePrime(keySize)
        q = generatePrime(keySize)

        modulus = p * q
        t = totient(modulus)

        publicKey = generateE(t)
        privateKey = calculateD(publicKey, t)
    }

    fun setClientPublicKey(clientPublicKey: ULong) {
        this.clientPublicKey = clientPublicKey
    }

    fun setClientModulus(clientModulus: ULong) {
        this.clientModulus = clientModulus
    }

    fun getPublicKey(): ULong {
        return publicKey
    }

    fun getModulus(): ULong {
        return modulus
    }

    fun getPrivateKey(): ULong {
        return privateKey
    }

    fun encrypt(m: String): List<Long> {
        if (clientPublicKey == 0UL || clientModulus == 0UL) {
            throw RuntimeException("Client public key or modulus not set")
        }

        val encryptedMessage = mutableListOf<Long>()

        for (c in m) {
            val encryptedChar = modPow(c.code.toLong().toULong(), clientPublicKey, clientModulus)
            encryptedMessage.add(encryptedChar)
        }

        return encryptedMessage
    }

    fun encryptToString(m: String): String {
        // Encrypt to "[1, 2, 3, 4]"
        val encryptedList = encrypt(m)
        return encryptedList.joinToString(separator = ", ", prefix = "[", postfix = "]")
    }

    private fun decrypt(encrypted: List<ULong>): String {
        val decryptedMessage = StringBuilder()

        for (encryptedChar in encrypted) {
            val decryptedChar = modPow(encryptedChar, privateKey, modulus).toChar()
            decryptedMessage.append(decryptedChar)
        }

        return decryptedMessage.toString()
    }

    fun decrypt(encrypted: String): String {
        // Remove brackets and split by comma and space
        val encryptedList = encrypted.substring(1, encrypted.length - 1).split(", ").map { it.toULong() }
        return decrypt(encryptedList)
    }


    private fun generatePrime(bigNum: Int): ULong {
        while (true) {
            val number = (bigNum / 2..bigNum).random().toULong()

            if (isPrime(number)) {
                return number
            }
        }
    }

    private fun totient(n: ULong): ULong {
        var result = n
        var i = 2UL
        var n = n

        while (i * i <= n) {
            if (n % i == 0UL) {
                result -= result / i
                while (n % i == 0UL) {
                    n /= i
                }
            }
            i += 1UL
        }

        if (n > 1UL) {
            result -= result / n
        }

        return result
    }

    private fun isPrime(n: ULong): Boolean {
        if (n < 2UL) return false

        for (i in 2UL..sqrt(n.toDouble()).toULong() + 1UL) {
            if (n % i == 0UL) return false
        }

        return true
    }

    private fun generateE(num: ULong): ULong {
        for (e in 2UL..<num) {
            if (gcd(e, num) == 1UL) {
                return e
            }
        }

        throw RuntimeException("Couldn't generate e")
    }

    private fun modPow(base: ULong, exponent: ULong, modulus: ULong): Long {
        var result = 1UL
        var exp = exponent
        var baseVar = base

        while (exp > 0UL) {
            if (exp % 2UL == 1UL) {
                result = (result * baseVar) % modulus
            }
            exp /= 2UL
            baseVar = (baseVar * baseVar) % modulus
        }

        return result.toLong()
    }

    private fun calculateD(e: ULong, totient: ULong): ULong {
        var k = 1UL

        while ((k * totient + 1UL) % e != 0UL) {
            k += 1UL
        }

        return (k * totient + 1UL) / e
    }


    private fun gcd(a: ULong, b: ULong): ULong {
        var x = a
        var y = b
        while (y != 0UL) {
            val temp = y
            y = x % y
            x = temp
        }
        return x
    }

    fun foramtForSending(): String {
        return "$publicKey::::${modulus}"
    }
}
